File: examples/multi_backend.py
"""
Example: Multi-Backend Telemetry
================================

This example shows how to send telemetry to multiple backends simultaneously.
Useful for:
- Development (console) + Production (Splunk)
- A/B testing different observability platforms
- Redundancy across backends

Prerequisites:
    pip install genai-telemetry

Usage:
    python multi_backend.py
"""

from genai_telemetry import (
    setup_telemetry,
    trace_llm,
    SplunkHECExporter,
    ElasticsearchExporter,
    ConsoleExporter,
    FileExporter,
    MultiExporter,
)


def example_1_simple_multi_backend():
    """Use the simple list-based configuration."""
    print("\n" + "=" * 60)
    print("Example 1: Simple Multi-Backend Configuration")
    print("=" * 60)
    
    # Configure multiple backends using the list syntax
    setup_telemetry(
        workflow_name="multi-backend-app",
        exporter=[
            {"type": "console", "colored": True},
            {"type": "file", "path": "/tmp/traces.jsonl"},
            # Uncomment for production:
            # {"type": "splunk", "url": "https://splunk:8088", "token": "xxx"},
            # {"type": "elasticsearch", "hosts": ["http://es:9200"]},
        ]
    )
    
    @trace_llm(model_name="gpt-4o", model_provider="openai")
    def mock_chat(message: str):
        # Mock response
        class MockResponse:
            class Usage:
                prompt_tokens = 100
                completion_tokens = 50
            usage = Usage()
        return MockResponse()
    
    mock_chat("Hello!")
    print("✓ Telemetry sent to console and file")


def example_2_custom_exporters():
    """Use custom exporter instances for more control."""
    print("\n" + "=" * 60)
    print("Example 2: Custom Exporter Instances")
    print("=" * 60)
    
    # Create custom exporter instances
    console_exporter = ConsoleExporter(colored=True, verbose=False)
    file_exporter = FileExporter(file_path="/tmp/custom_traces.jsonl")
    
    # Combine them
    multi_exporter = MultiExporter([
        console_exporter,
        file_exporter,
    ])
    
    # Use the multi-exporter directly
    setup_telemetry(
        workflow_name="custom-exporters-app",
        exporter=multi_exporter
    )
    
    @trace_llm(model_name="claude-3-sonnet", model_provider="anthropic")
    def mock_anthropic_chat(message: str):
        class MockResponse:
            class Usage:
                input_tokens = 75
                output_tokens = 125
            usage = Usage()
        return MockResponse()
    
    mock_anthropic_chat("Hello from Anthropic!")
    print("✓ Telemetry sent via custom exporters")


def example_3_console_plus_file_shorthand():
    """Use the console=True and file_path shorthand."""
    print("\n" + "=" * 60)
    print("Example 3: Console + File Shorthand")
    print("=" * 60)
    
    # Quick shorthand for console + file
    setup_telemetry(
        workflow_name="shorthand-app",
        exporter="console",  # Primary exporter
        file_path="/tmp/shorthand_traces.jsonl"  # Also write to file
    )
    
    @trace_llm(model_name="gpt-4o-mini", model_provider="openai")
    def quick_test(message: str):
        class MockResponse:
            class Usage:
                prompt_tokens = 25
                completion_tokens = 10
            usage = Usage()
        return MockResponse()
    
    quick_test("Quick test!")
    print("✓ Telemetry sent via shorthand config")


if __name__ == "__main__":
    print("Multi-Backend Telemetry Examples")
    print("=" * 60)
    
    example_1_simple_multi_backend()
    example_2_custom_exporters()
    example_3_console_plus_file_shorthand()
    
    print("\n" + "=" * 60)
    print("All examples completed!")
    print("Check /tmp/traces.jsonl for file output")
    print("=" * 60)

File: js/examples/multi_backend.ts
/**
 * Multi-Backend Example
 * 
 * This example demonstrates how to send telemetry to multiple backends
 * simultaneously - useful for migration scenarios or redundancy.
 */

import { setupTelemetry, traceLLM, getTelemetry } from 'genai-telemetry';

// Initialize telemetry with multiple exporters
setupTelemetry({
  workflowName: 'multi-backend-app',
  exporter: [
    // Send to Splunk for real-time dashboards
    {
      type: 'splunk',
      url: process.env.SPLUNK_HEC_URL || 'http://localhost:8088',
      token: process.env.SPLUNK_HEC_TOKEN || 'test-token',
      index: 'genai_traces'
    },
    // Send to Elasticsearch for long-term storage
    {
      type: 'elasticsearch',
      hosts: [process.env.ES_HOST || 'http://localhost:9200'],
      index: 'genai-traces'
    },
    // Send to OTLP collector (can forward to Jaeger, Datadog, etc.)
    {
      type: 'otlp',
      endpoint: process.env.OTLP_ENDPOINT || 'http://localhost:4318'
    },
    // Also log to console for development
    {
      type: 'console',
      colored: true,
      verbose: false
    }
  ],
  batchSize: 10,
  flushInterval: 5
});

// Simulated LLM call
const chat = traceLLM(
  async (message: string) => {
    // Simulate API delay
    await new Promise(resolve => setTimeout(resolve, 100));
    
    return {
      choices: [{
        message: { content: `Response to: ${message}` }
      }],
      usage: {
        prompt_tokens: message.split(' ').length * 2,
        completion_tokens: 50
      }
    };
  },
  { modelName: 'gpt-4o', modelProvider: 'openai' }
);

async function main() {
  console.log('Multi-Backend Telemetry Example');
  console.log('================================\n');
  console.log('Sending traces to:');
  console.log('  - Splunk HEC');
  console.log('  - Elasticsearch');
  console.log('  - OTLP Collector');
  console.log('  - Console\n');

  // Make several calls
  for (let i = 1; i <= 5; i++) {
    console.log(`Making call ${i}...`);
    await chat(`This is test message number ${i}`);
  }

  // Manually send a custom span
  const telemetry = getTelemetry();
  await telemetry.sendSpan('TOOL', 'custom-operation', {
    durationMs: 42,
    toolName: 'example-tool',
    customField: 'custom-value'
  });

  console.log('\nDone! Traces sent to all configured backends.');
}

main().catch(console.error);
